### PR TITLE
[internal] adds tests for internal files

### LINKED ISSUE #149: Do we need tests for the internal and cmd dirs?
```
$ go test ./...
ok  	go.yaml.in/yaml/v4	2.286s
?   	go.yaml.in/yaml/v4/cmd/go-yaml	[no test files]
?   	go.yaml.in/yaml/v4/internal/libyaml	[no test files]
ok  	go.yaml.in/yaml/v4/internal/testutil/assert	(cached)
ok  	go.yaml.in/yaml/v4/yts	0.099s
```

Also should make test trigger `go test ./...` instead of just `go test`?

### PR DESCRIPTION
**Pull request by colinjlacy**
This branch adds unit tests in the integration
style to the internal/libyaml directory. it
focuses primarily on the majority use cases and
does not get too creative with edge-case yaml
examples, which can likely be filled in as issues
are opened and bugs are squashed. The majority of
these tests were written by Claude in cursor.

To test, simply run:
```
make test-all
go test ./... --cover
```

fixes #149 in part

### CODE REVIEW

**Comment by ingydotnet** [comment]
> ```
> go test ./... --cover
> ```

Can you add `test-cover` rule to the Makefile for that?

File: internal/libyaml/reader_test.go [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
```
Comment (ccoVeille) [inline internal/libyaml/reader_test.go]: I'm unsure we need to apply the legacy license to file we add.

They do use old code that were licensed under these Licenses

But here, these are files we add for the yaml/go-yaml project.

I wouldn't use a license header, or if I do, I would clearly state these new files are licensed under the license of the yaml/go-yaml project.

So I would simply use

```suggestion
// SPDX-License-Identifier: Apache-2.0
```


What do you think @ingydotnet ?

File: internal/libyaml/reader_test.go [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+	}
+	if parser.Problem != "test problem" {
+		t.Errorf("setReaderError() Problem = %q, want \"test problem\"", parser.Problem)
+	}
+	if parser.ProblemOffset != 10 {
+		t.Errorf("setReaderError() ProblemOffset = %d, want 10", parser.ProblemOffset)
+	}
+	if parser.ProblemValue != 0x1234 {
+		t.Errorf("setReaderError() ProblemValue = %#x, want 0x1234", parser.ProblemValue)
+	}
```
Comment (ccoVeille) [inline internal/libyaml/reader_test.go]: Here and everywhere, you could have used the assertion package we created

https://pkg.go.dev/go.yaml.in/yaml/v4@v4.0.0-rc.3/internal/testutil/assert

But why not

File: internal/libyaml/reader_test.go [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+func TestYamlStringReadHandler(t *testing.T) {
+	parser := NewParser()
+	input := []byte("test data")
+	parser.input = input
+	parser.input_pos = 0
+
+	buffer := make([]byte, 10)
+	n, err := yamlStringReadHandler(&parser, buffer)
+
+	if err != nil && err != io.EOF {
```
Comment (ccoVeille) [inline internal/libyaml/reader_test.go]: Here and everywhere `errors.Is` is the way to check error equality

File: GNUmakefile [hunk 4]
```diff
@@ -68,6 +68,9 @@ count ?= 1
 test: $(GO-DEPS)
 	go test$(if $v, -v) -vet=off .
 
+test-cover: $(GO-DEPS)
+	go test$(if $v, -v) ./... --cover
+
```
Comment (ccoVeille) [inline GNUmakefile:70]: I know @ingydotnet asked you to add this, but I would have simply used this

```suggestion
test: $(GO-DEPS)
go test$(if $v, -v) -vet=off --cover . 

```

Because the cover is fine in the context of the test, here you are adding cover for everything, while we stated we cannot use ./...

If there is a need for having a test-cover rule, you should use this

```suggestion
test: $(GO-DEPS)
go test$(if $v, -v) -vet=off .

test-cover: $(GO-DEPS)
	go test$(if $v, -v) -vet=off --cover .

```

File: internal/libyaml/reader_test.go (hunk 1 above)
Comment (ingydotnet) [reply internal/libyaml/reader_test.go]: I agree, it doesn't belong.

File: GNUmakefile (hunk 4 above)
Comment (colinjlacy) [reply GNUmakefile:70]: @ccoVeille sorry, I didn't know we aren't using `./...`. Do you mind explaining why that is?

Comment (ccoVeille) [reply GNUmakefile:70]: It's because the existence of yaml test suite in yts.

You can find some context here:

https://github.com/yaml/go-yaml/pull/171#discussion_r2487945015

I would like to suggest you to read all resolved comments, including thr one opened by Copilot.

We tried to fix things with #171 #172, #173 and #174

But there was no agreement, so #175 was opened.

I would say we have people who use `make test`command to launch tests, and Gophers with muscle memories who use `go test ./...`

It's uneasy to make everyone happy 😅

So for now, there is no real progress on the situation.

I would love getting feedbacks from you here.

Comment (ccoVeille) [reply GNUmakefile:70]: But for now, using this could be a solution for your PR

```suggestion
test: $(GO-DEPS)
go test$(if $v, -v) -vet=off --cover . internal/...

```

Comment (colinjlacy) [reply GNUmakefile:70]: Thanks for the context. I can see this has been a back-and-forth issue. I'm an old Go hand as well, so the muscle memory is real for me too. Still, I understand the other perspective entirely. I added the modified line as you suggested and we can adjust from there if/when more tests are added.

File: GNUmakefile [hunk 5]
```diff
@@ -68,6 +68,9 @@ count ?= 1
 test: $(GO-DEPS)
 	go test$(if $v, -v) -vet=off .
 
+test-cover: $(GO-DEPS)
+	go test$(if $v, -v) -vet=off --cover . ./internal/...
```
Comment (ccoVeille) [inline GNUmakefile:69]: Let's continue there

If you take a look at my previous comment, you could see I was suggesting this


```suggestion
test: $(GO-DEPS)
	go test$(if $v, -v) -vet=off --cover . ./internal/...
```

But if there is real will to have a separate `make test-cover` this should be used.

```suggestion
test: $(GO-DEPS)
	go test$(if $v, -v) -vet=off . ./internal/...
 
test-cover: $(GO-DEPS)
	go test$(if $v, -v) -vet=off --cover . ./internal/...
```

**Review by ccoVeille — APPROVED** [review]
I didn't review all tests. But based in the fact you are not updating the code, but adding tests I feel like I can approve anyway.

**Comment by colinjlacy** [comment]
Closing in favor of #194
